hazelcast/src/main/java/com/hazelcast/config/ConfigReplacerHelper.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+import com.hazelcast.config.replacer.spi.ConfigReplacer;
+import com.hazelcast.config.yaml.ElementAdapter;
+import com.hazelcast.internal.yaml.MutableYamlNode;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+import java.util.List;
+
+import static java.lang.String.format;
+
+/**
+ * Helper class for replacing variables in the XML and YAML configuration DOMs
+ */
+final class ConfigReplacerHelper {
+
+    private static final ILogger LOGGER = Logger.getLogger(ConfigReplacerHelper.class);
+
+    private ConfigReplacerHelper() {
+    }
+
+    static void traverseChildrenAndReplaceVariables(Node root, List<ConfigReplacer> replacers, boolean failFast) {
+
+        // Use all the replacers on the content
+        for (ConfigReplacer replacer : replacers) {
+            traverseChildrenAndReplaceVariables(root, replacer, failFast);
+        }
+    }
+
+    private static void traverseChildrenAndReplaceVariables(Node root, ConfigReplacer replacer, boolean failFast) {
+        NamedNodeMap attributes = root.getAttributes();
+        if (attributes != null) {
+            for (int k = 0; k < attributes.getLength(); k++) {
+                Node attribute = attributes.item(k);
+                replaceVariables(attribute, replacer, failFast);
+            }
+        }
+        if (root.getNodeValue() != null) {
+            replaceVariables(root, replacer, failFast);
+        }
+        final NodeList childNodes = root.getChildNodes();
+        for (int k = 0; k < childNodes.getLength(); k++) {
+            Node child = childNodes.item(k);
+            if (child != null) {
+                traverseChildrenAndReplaceVariables(child, replacer, failFast);
+            }
+        }
+    }
+
+    private static void replaceVariables(Node node, ConfigReplacer replacer, boolean failFast) {
+        if (node == null) {
```

Comment:
Do we really need this check? All paths leading here are passing non-null values.

Reply:
This is similar to the above one.

Translated this:
```
        String xml = HAZELCAST_START_TAG
                + "   <map name=\"people\">\n"
                + "       <attributes>\n"
                + "           <attribute extractor=\"\">weight</attribute>\n"
                + "       </attributes>"
                + "   </map>"
                + HAZELCAST_END_TAG;
```
to this:
```
        String yaml = ""
                + "hazelcast:\n"
                + "  map:\n"
                + "    people:\n"
                + "      attributes:\n"
                + "        weight:\n"
                + "          extractor: \n";
```
Notice the missing node at the end after the extractor. Added empty string here as well.